crates/schema/src/error.rs

```diff
@@ -0,0 +1,189 @@
+pub type ValidationErrors = ErrorStream<ValidationError>;
+
+/// A single validation error.
+///
+/// Many variants of this enum store `RawIdentifier`s rather than `Identifier`s.
+/// This is because we want to support reporting errors about module entities with invalid names.
+#[derive(thiserror::Error, Debug, PartialOrd, Ord, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum ValidationError {
+    #[error("name `{name}` is used for multiple entities")]
+    DuplicateName { name: Identifier }, // use an Identifier because this is checked for validated names.
+    #[error("module contains invalid identifier: {error}")]
+    IdentifierError { error: IdentifierError },
+    #[error("table `{}` has unnamed column `{}`, which is forbidden.", column.table, column.column)]
+    UnnamedColumn { column: RawColumnName },
+    #[error("type `{type_name}` is not annotated with a custom ordering, but uses one: {bad_type:?}")]
+    TypeHasIncorrectOrdering {
+        type_name: RawIdentifier,
+        ref_: AlgebraicTypeRef,
+        /// Could be a sum or product.
+        bad_type: AlgebraicType,
+    },
+    #[error("column `{column}` referenced by def {def} not found in table `{table}`")]
+    ColumnNotFound {
+        table: RawIdentifier,
+        def: RawIdentifier,
+        column: ColId,
+    },
+    #[error("column `{column}` has `Constraints::unset()`")]
+    ConstraintUnset { column: RawColumnName, name: RawIdentifier },
+    #[error("Attempt to define a column `{column}` with more than 1 auto_inc sequence")]
+    OneAutoInc { column: RawColumnName },
+    #[error("Only Btree Indexes are supported: index `{index}` is not a btree")]
+    OnlyBtree { index: RawIdentifier },
+    #[error("def `{def}` has duplicate columns: {columns:?}")]
+    DuplicateColumns { def: RawIdentifier, columns: ColList },
+    #[error("invalid sequence column type: `{column}` with type `{column_type:?}` in sequence `{sequence}`")]
+    InvalidSequenceColumnType {
+        sequence: RawIdentifier,
+        column: RawColumnName,
+        column_type: AlgebraicType,
+    },
+    #[error("invalid sequence range information: expected {min_value:?} <= {start:?} <= {max_value:?} in sequence `{sequence}`")]
+    InvalidSequenceRange {
+        sequence: RawIdentifier,
+        min_value: Option<i128>,
+        start: Option<i128>,
+        max_value: Option<i128>,
+    },
+    #[error("Table {table} has invalid product_type_ref {ref_}")]
+    InvalidProductTypeRef {
+        table: RawIdentifier,
+        ref_: AlgebraicTypeRef,
+    },
+    #[error("Type {type_name} has invalid ref: {ref_}")]
+    InvalidTypeRef {
+        type_name: RawIdentifier,
+        ref_: AlgebraicTypeRef,
+    },
+    #[error("A scheduled table must have columns `scheduled_id: u64` and `scheduled_at: ScheduledAt`, but table `{table}` has columns {columns:?}")]
+    ScheduledIncorrectColumns { table: RawIdentifier, columns: ProductType },
+    #[error("{location} is not in nominal normal form: {ty:?}")]
```

Comment:
Yeah it's not great.
This error should never happen to users using the official SDKs though, it would only happen if the SDK is generating malformed ModuleDefs. Nothing the user can do should be able to cause this, it's really an error for SDK developers.